### ISSUE TITLE
Suggestion: Use string literals instead of named routes or URL helpers in tests

### ISSUE THREAD

**Issue by ohbarye** [rubocop/rails-style-guide#328]
Hi. This is a proposal about named routes usage in tests. I'd like to hear what you think.

### Proposal

Use string literals instead of named routes or URL helpers in tests.

```ruby
# bad
get photos_path
get photo_path(id)
get edit_photo_path(id)
```

```ruby
# good
get "/photos"
get "/photos/#{id}"
get "/photos/#{id}/edit"
```

and so on.

## Reason

Integration test (e.g. request specs, system specs) is to mimic user eyes and actions. A user could see "/photos", but not "photos_path". I agree with the following idea.

>The more your tests resemble the way your software is used, the more confidence they can give you.
>https://twitter.com/kentcdodds/status/977018512689455106

From a programmer's view, 
- the bad pattern slightly increases cognitive load that forces me to build actual URLs in my mind.
- the bad pattern sometimes discourages Test Driven Development.
    - In TDD, a test comes first. At this point, I know actual URL but I don't know the named routes because the `routes.rb` has not been edited yet. (=I don't care about named routes at that moment).

## Possible Objection

named routes possibly have merits in some points.

- You don't have to edit test cases when you edit `routes.rb`. It automatically follows up.
- It could be consistent to use named routes in tests if you use named routes in production code.

## References

- https://guides.rubyonrails.org/routing.html

**Comment by pirj** [comment]
There's some controversy there in the Rails Testing Rails Applications guide.

[Integration](https://guides.rubyonrails.org/testing.html#implementing-an-integration-test):
```ruby
class BlogFlowTest < ActionDispatch::IntegrationTest
  test "can see the welcome page" do
    get "/"
```

[System](https://guides.rubyonrails.org/testing.html#implementing-a-system-test):
```ruby
class ArticlesTest < ApplicationSystemTestCase
  test "viewing the index" do
    visit articles_path
```

Wondering why is it so.

If you don't write route tests and use URL helpers in system/integration, the mapping is not covered. It is risky to have a green CI after changing routes.

**Comment by RoshanNair22** [comment]
I agree with all the responses. But we all have our own way of solving our problems, right?
Yeah, so get on with it.

Also, FYI, if any of our curious readers want to know more about [test driven development](https://thinkpalm.com/blogs/what-is-test-driven-development-and-what-are-its-benefits/), please read the blog written my colluagues.

**Comment by moznion** [comment]
I completely agree with this suggestion. Has there been any progress on it?

**Comment by pirj** [comment]
Would you like to send a pr to kick start tims, @moznion ?

**Comment by moznion** [comment]
@pirj Thank you for your suggestion. Why not, I'll work on it.

**Comment by moznion** [comment]
@pirj
Hi, I just sent a pull request to this repo #359, and the actual cop implementation is here (not yet pull-requested): https://github.com/rubocop/rubocop-rails/compare/master...moznion:rubocop-rails:http-url?expand=1

What do you think about them?

**Comment by andyw8** [comment]
> There's some controversy there in the Rails Testing Rails Applications guide.
> 
> [Integration](https://guides.rubyonrails.org/testing.html#implementing-an-integration-test):
> 
> ```ruby
> class BlogFlowTest < ActionDispatch::IntegrationTest
>   test "can see the welcome page" do
>     get "/"
> ```
> 
> [System](https://guides.rubyonrails.org/testing.html#implementing-a-system-test):
> 
> ```ruby
> class ArticlesTest < ApplicationSystemTestCase
>   test "viewing the index" do
>     visit articles_path
> ```
> 
> Wondering why is it so.
> 
> If you don't write route tests and use URL helpers in system/integration, the mapping is not covered. It is risky to have a green CI after changing routes.

Agreed. I think it would be best to clarify the Rails docs first, then add some guidance here.

**Comment by AriseshineSky** [comment]
- [ ] j jij
